### PR TITLE
Harmonize gRPC calls

### PR DESCRIPTION
**Pull request by garuppel**
Create handle-functions for gRPC calls
- orchestrator.StoreAssessmentResult(s)
- evidenceStore.StoreEvidence(s)
- assessment.AssessEvidence(s)

### CODE REVIEW

File: service/evidence/evidence_store.go [hunk 1]
```diff
@@ -36,63 +36,66 @@ func init() {
...
-			for _, hook := range s.EvidenceHook {
-				go hook(nil, newError)
-			}
+	if err != nil {
+		resp = &evidence.StoreEvidenceResponse{
+			Status: false,
 		}
 
-		return resp, status.Errorf(codes.InvalidArgument, "invalid req: %v", err)
+		return resp, status.Errorf(codes.Internal, "Error while handling evidence: %v", err)
```
Comment (oxisto) [inline service/evidence/evidence_store.go:48]: This is probably the correct PR for this: The problem that I see here is that it is now indistinguishable whether we have an internal error (such as the DB connection was lost or so), which the user cannot fix from an invalid argument error, which the user can fix, e.g. by supplying the correct argument.

File: service/assessment/assessment.go [hunk 2]
```diff
@@ -52,9 +52,9 @@ Service is an implementation of the Clouditor Assessment service. It should not
 NewService constructor should be used. It implements the AssessmentServer interface
 */
 type Service struct {
-	// ResultHook is a hook function that can be used if one wants to be
+	// resultHook is a hook function that can be used if one wants to be
 	// informed about each assessment result
-	ResultHook []func(result *assessment.AssessmentResult, err error)
+	resultHook []func(result *assessment.AssessmentResult, err error)
```
Comment (oxisto) [inline service/assessment/assessment.go:57]: Let's call it resultHooks. I would also suggest declaring a `ResultHookFunc` type, to make it easier for users which function signature to implement.

File: service/evidence/evidence_store.go [hunk 3]
```diff
@@ -20,7 +20,7 @@ type Service struct {
 	evidences map[string]*evidence.Evidence
 
 	// Hook
-	EvidenceHook []func(result *evidence.Evidence, err error)
+	evidenceHook []func(result *evidence.Evidence, err error)
```
Comment (oxisto) [inline service/evidence/evidence_store.go:23]: evidenceHooks?

File: service/evidence/evidence_store.go (hunk 1 above)
Comment (oxisto) [reply service/evidence/evidence_store.go:48]: So I would suggest to at least call the validate function outside of the handleXX function. All errors from validate should be invalid arguments. all errors from the handleXXX can be internal errors.

File: service/assessment/assessment.go [hunk 4]
```diff
@@ -52,9 +52,9 @@ Service is an implementation of the Clouditor Assessment service. It should not
 NewService constructor should be used. It implements the AssessmentServer interface
 */
 type Service struct {
-	// ResultHook is a hook function that can be used if one wants to be
+	// resultHook is a hook function that can be used if one wants to be
```
[the remark this answers is not in the document]
Comment (lebogg) [reply service/assessment/assessment.go:55]: and maybe plural: `resultHooks`?

File: service/assessment/assessment_test.go [hunk 5]
```diff
@@ -191,7 +191,7 @@ func TestService_AssessEvidences(t *testing.T) {
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			s := Service{
-				ResultHook:                    tt.fields.ResultHook,
+				resultHook:                    tt.fields.ResultHook,
```
Comment (lebogg) [inline service/assessment/assessment_test.go:194]: Would be also plural then (`fields.resultHooks`)

File: service/discovery/discovery_test.go [hunk 6]
```diff
@@ -337,7 +337,7 @@ func (mockIsCloudResource) GetCreationTime() *time.Time {
 //	var ready = make(chan bool)
 //
 //	assessmentServer := standalone.NewAssessmentServer().(*service_assessment.Service)
-//	assessmentServer.ResultHook = func(result *assessment.Result, err error) {
+//	assessmentServer.resultHook = func(result *assessment.Result, err error) {
```
Comment (lebogg) [inline service/discovery/discovery_test.go:340]: Much commented out code here. Maybe remove or stage it?

File: service/evidence/evidence_store.go (hunk 3 above)
Comment (lebogg) [inline service/evidence/evidence_store.go:23]: Also plural?

File: service/assessment/assessment.go [hunk 7]
```diff
@@ -171,17 +163,21 @@ func (s Service) handleEvidence(evidence *evidence.Evidence) error {
...
-				go hook(result, nil)
-			}
-		}
+		s.informHook(result, nil)
 	}
 
 	return nil
 }
 
+func (s Service) informHook(result *assessment.AssessmentResult, err error) {
```
Comment (lebogg) [inline service/assessment/assessment.go:172]: informHooks?

File: service/orchestrator/orchestrator.go [hunk 8]
```diff
@@ -213,51 +213,56 @@ func (s *Service) StoreAssessmentResults(stream orchestrator.Orchestrator_StoreA
...
 
-	return nil
+	return
 }
 
 func (s *Service) RegisterAssessmentResultHook(hook func(result *assessment.AssessmentResult, err error)) {
 	s.AssessmentResultsHook = append(s.AssessmentResultsHook, hook)
 }
 
+func (s Service) informHook(result *assessment.AssessmentResult, err error) {
```
Comment (lebogg) [inline service/orchestrator/orchestrator.go:257]: informHooks?

File: service/assessment/assessment.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (garuppel) [reply service/assessment/assessment.go:55]: :+1:

File: service/discovery/discovery_test.go (hunk 6 above)
Comment (garuppel) [reply service/discovery/discovery_test.go:340]: I didn't comment it out, but I think we do not need it any more? @oxisto @lebogg

Comment (oxisto) [reply service/discovery/discovery_test.go]: Yeah we can remove it

File: service/evidence/evidence_store.go (hunk 1 above)
Comment (garuppel) [reply service/evidence/evidence_store.go]: @oxisto I changed it. But I think it it sufficient if the StoreEvidences() calls the StoreEvidence(). We don't need the handle method. Please, can you take a quick look: https://github.com/clouditor/clouditor/blob/8e6f24de89f83fca5b573138a8f24edb41ea5e60/service/evidence/evidence_store.go#L39

Comment (garuppel) [reply service/evidence/evidence_store.go]: Should we than do the same for orchestrator.StoreAssessmentResult(s)()? There is the same error regarding the status message Internal vs InvalidArgument.

**Comment by garuppel** [comment]
I've changed the assessment.AssessEvidence(s)() and the orchestratore.StoreAssessmentResult(s)() as well. Now, the stream method calls the unary method.

**Review by lebogg — APPROVED** [review]

**Pull request merged**